src/pybamm/entry_points.py

```diff
@@ -0,0 +1,144 @@
+            EntryPoint._instances += 1
+            self._all_entries = dict()
+            self.group = group
+            for entry_point in self.get_entries(self.group):
+                self._all_entries[entry_point.name] = entry_point
+
+    @staticmethod
+    def get_entries(group_name):
+        """Wrapper for the importlib version logic"""
+        if sys.version_info < (3, 10):  # pragma: no cover
+            return importlib.metadata.entry_points()[group_name]
+        else:
+            return importlib.metadata.entry_points(group=group_name)
+
+    def __new__(cls, group):
+        """Ensure only two instances of entry points exist, one for parameter sets and the other for models"""
+        if EntryPoint._instances < 2:
+            cls.instance = super().__new__(cls)
+        return cls.instance
+
+    def __getitem__(self, key) -> dict:
+        return self._load_entry_point(key)()
+
+    def _load_entry_point(self, key) -> Callable:
+        """Check that ``key`` is a registered ``parameter_sets`` or ``models` ,
+        and return the entry point for the parameter set/model, loading it needed."""
+        if key not in self._all_entries:
+            raise KeyError(f"Unknown parameter set or model: {key}")
+        ps = self._all_entries[key]
+        try:
+            ps = self._all_entries[key] = ps.load()
+        except AttributeError:
+            pass
+        return ps
+
+    def __iter__(self):
+        return self._all_entries.__iter__()
+
+    def __len__(self) -> int:
+        return len(self._all_entries)
+
+    def get_docstring(self, key):
+        """Return the docstring for the ``key`` parameter set or model"""
+        return textwrap.dedent(self._load_entry_point(key).__doc__)
+
+    def __getattribute__(self, name):
+        try:
+            # For backwards compatibility, parameter sets that used to be defined in
+            # this file now return the name as a string, which will load the same
+            # parameter set as before when passed to `ParameterValues`
+            # Bypass the overloaded __getitem__ and __iter__ to avoid recursion
+            _all_entries = super().__getattribute__("_all_entries")
+            if name in _all_entries:
+                msg = (
+                    f"Parameter sets should be called directly by their name ({name}), "
+                    f"instead of via pybamm.parameter_sets (pybamm.parameter_sets.{name})."
+                )
+                warnings.warn(msg, DeprecationWarning, stacklevel=2)
+                return name
+        except AttributeError:
+            pass  # Handle the attribute error normally
+
+        return super().__getattribute__(name)
+
+
+#: Singleton Instance of :class:EntryPoint initialised with pybamm_parameter_sets"""
+parameter_sets = EntryPoint(group="pybamm_parameter_sets")
+
+#: Singleton Instance of :class:EntryPoint initialised with pybamm_models"""
+model_sets = EntryPoint(group="pybamm_models")
```

Comment:
this should just be called `models`, a `parameter_set` is a set of parameters (and a model is a set of equations), there is no set of models

Reply:
Yes that was intentional, `models` wasn't working because the `pybamm.models` namespace is already taken by the collection of battery models which gets overridden first. We could either change the variable name which was my temporary fix (`model_sets`) to something more sensible or we could change the namespace for the entry point models to something like `pybamm.entrypoint.models`. I'd prefer the first option though if you could come up with a sensible variable name for the models instance because I have a limited knowledge with the battery library.

Reply:
Understood! In order to keep it consistent with parameter sets let's keep it as is then

Reply:
Sorry for the late response. I have abstracted the `models` instance inside the `dispatch` namespace as @agriyakhetarpal mentioned. I did not move the `parameter_sets` instance inside the dispatch namespace, as it would be a breaking change and also fail a lot of tests because `pybamm.parameter_sets` wouldn't work instead, people would have to use `pybamm.dispatch.parameter_sets`, though both of these work as of now. The `models` instance (entry point) can now be accessed through `pybamm.dispatch.models` and `list(pybamm.dispatch.models)` would list all the battery models available through the entry point. To load the model, `pybamm.Model("SPM")` would return the model object.